File: problems/0026-remove-duplicates-from-sorted-array/solution.py
"""
Problem Number: 26
Problem Name: Remove Duplicates from Sorted Array
Difficulty: Easy
Topics: Array, Two Pointers
Companies: Facebook, Microsoft, Google
"""

from typing import List

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        """
        Approach: Two Pointers
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        if not nums:
            return 0
        
        i = 0
        for j in range(1, len(nums)):
            if nums[j] != nums[i]:
                i += 1
                nums[i] = nums[j]
                
        return i + 1

# Test cases
if __name__ == "__main__":
    solution = Solution()
    nums1 = [1,1,2]
    k1 = solution.removeDuplicates(nums1)
    print(k1, nums1[:k1]) # 2, [1,2]

    nums2 = [0,0,1,1,1,2,2,3,3,4]
    k2 = solution.removeDuplicates(nums2)
    print(k2, nums2[:k2]) # 5, [0,1,2,3,4]

File: problems/0026-remove-duplicates-from-sorted-array/solution.js
/**
 * @problemNumber: 26
 * @problemName: Remove Duplicates from Sorted Array
 * @difficulty: Easy
 * @topics: Array, Two Pointers
 * @companies: Facebook, Microsoft, Google
 */

/**
 * Approach: Two Pointers
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 *
 * @param {number[]} nums
 * @return {number}
 */
function removeDuplicates(nums) {
    if (nums.length === 0) return 0;

    let i = 0;
    for (let j = 1; j < nums.length; j++) {
        if (nums[j] !== nums[i]) {
            i++;
            nums[i] = nums[j];
        }
    }

    return i + 1;
}

// Test cases
let nums1 = [1, 1, 2];
let k1 = removeDuplicates(nums1);
console.log(k1, nums1.slice(0, k1)); // 2, [1,2]

let nums2 = [0, 0, 1, 1, 1, 2, 2, 3, 3, 4];
let k2 = removeDuplicates(nums2);
console.log(k2, nums2.slice(0, k2)); // 5, [0,1,2,3,4]

module.exports = { removeDuplicates };
